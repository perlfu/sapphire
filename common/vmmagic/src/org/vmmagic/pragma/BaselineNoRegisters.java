/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * Some methods shouldn't use registers for locals and stack values.
 * E.g. CollectorThread can't relocate refs in registers of its own.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Pragma
public @interface BaselineNoRegisters { }
