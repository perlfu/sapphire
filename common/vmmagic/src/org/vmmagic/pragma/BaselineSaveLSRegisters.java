/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * Methods with this pragma that are BaselineCompiled should save in its prologue, ALL registers that
 * can be used to store local and stack registers in any BaselineCompiled method. Used by OSR.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Pragma
public @interface BaselineSaveLSRegisters { }
