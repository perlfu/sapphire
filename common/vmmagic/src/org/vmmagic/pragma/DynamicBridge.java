/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * Methods of a class that implements this interface are treated specially
 * by the compilers.
 *
 * <p>Instead of saving just the non-volatile registers used by the method into
 * the register save area of the method's stackframe, the compiler generates
 * code to save *all* GPR and FPR registers except GPR0, FPR0, JTOC, and FP.
 *
 * <p>Methods of a class that implement this interface may not return.
 *    (it is assumed that execution is resumed via a call to Magic.dynamicBridgeTo)
 *
 * @see org.jikesrvm.runtime.Magic#dynamicBridgeTo(org.jikesrvm.ArchitectureSpecific.CodeArray)
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
@Pragma
public @interface DynamicBridge { }
