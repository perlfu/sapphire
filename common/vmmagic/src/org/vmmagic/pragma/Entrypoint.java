/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * The entrypoint annotation indicates that the method or field is
 * directly accessed by the compiler. We cache resolved values for
 * these elements in Entrypoints.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.METHOD, ElementType.FIELD})
@Pragma
public @interface Entrypoint { }
