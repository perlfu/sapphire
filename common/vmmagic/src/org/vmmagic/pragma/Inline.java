/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * This pragma indicates that a particular method should always be inlined
 * by the optimizing compiler.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Pragma
public @interface Inline {
  /**
   * Enumeration defining when to inline
   */
  public enum When {
    /**
     * Always inline, regardless of arguments
     */
    Always,
    /**
     * Inline when all the arguments are constants
     */
    AllArgumentsAreConstant,
    /**
     * Inline when the specified arguments are constants
     */
    ArgumentsAreConstant,
    /**
     * Inline when the VM is built without Assertions (VM.VerifyAssertions == false).
     * Note: It would be nicer to have the more general ExpressionIsTrue annotation,
     * but the argument expression to the annotation is restricted to be a fairly
     * trivial constant, and that isn't enough to handle how VM.VERIFY_ASSERTIONS
     * is defined in MMTk.
     */
    AssertionsDisabled
  }
  /**
   * When to inline, default When.Always
   */
  When value() default When.Always;
  /**
   * Arguments that must be constant to cause inlining. NB for static methods 0
   * is the first argument whilst for virtual methods 0 is this
   */
  int[] arguments() default {};
}
