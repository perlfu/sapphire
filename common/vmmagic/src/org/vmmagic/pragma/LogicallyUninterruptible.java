/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * A pragma that can be used to declare that a particular method is logically
 * uninterruptible even though it contains bytecodes that are actually
 * interruptible.
 *
 * The effect of this pragma is to suppress warning messages about violations of
 * uninterruptiblity when compiling a method that throws this exception. There
 * are two cases in which using the pragma is justified.
 * <ul>
 * <li> Uninterruptibility is ensured via some other mechanism. For example, the
 * method explicitly disables threadswitching around the interruptible regions
 * (VM.sysWrite on String). Or the interruptible regions are not reachable when
 * the VM is running (various VM.sysWrite that check VM.runningVM).
 * <li> The interruptible regions represent an 'error' condition that will never
 * be executed unless the VM is already in the process of reporting an error,
 * for example RuntimeEntrypoints.raiseClassCastException.
 * <ul>
 * Extreme care must be exercised when using this pragma since it suppresses the
 * checking of uninterruptibility.
 * <p>
 * Use of this pragma is being phased out since it lumps together two possible
 * special cases. Use either {@link Unpreemptible} or
 * {@link UninterruptibleNoWarn} instead.
 * {@link <a href="http://jira.codehaus.org/browse/RVM-115">RVM-115</a>} for more
 * context.
 * @deprecated
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Pragma
public @interface LogicallyUninterruptible { }
