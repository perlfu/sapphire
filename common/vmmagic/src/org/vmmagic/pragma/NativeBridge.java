/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * Methods of a class that implements this interface are treated specially
 * by the compilers:
 * <ul>
 *  <li>They are only called from C or C++ program</li>
 *  <li>The compiler will generate the necessary prolog to insert a glue stack
 *   frame to map from the native stack/register convention to RVM's convention</li>
 *  <li>It is an error to call these methods from Java</li>
 * </ul>
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
@Pragma
public @interface NativeBridge { }
