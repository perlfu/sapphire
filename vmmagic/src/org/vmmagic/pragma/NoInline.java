/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.vmmagic.pragma;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.ElementType;
import org.vmmagic.Pragma;

/**
 * This pragma indicates that a particular method should never be inlined
 * by the optimizing compiler.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Pragma
public @interface NoInline { }
