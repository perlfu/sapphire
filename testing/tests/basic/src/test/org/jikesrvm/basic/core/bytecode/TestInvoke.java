/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.basic.core.bytecode;

import org.vmmagic.pragma.NoInline;

/*
*/
class TestInvoke {

  static interface MyInterface {
    void performMagic();
  }

  static class TypeA {
    TypeA() {System.out.println("TypeA.<init>()");}

    void f() { System.out.println("TypeA.f()"); }
  }

  static class TypeB extends TypeA {
    TypeB() {System.out.println("TypeB.<init>()");}

    //invokevirtual
    void f() { System.out.println("TypeB.f()"); }

    //invokestatic
    static int g(int value) { return 3 + value; }
  }

  static class TypeC extends TypeB implements MyInterface {
    TypeC() {System.out.println("TypeC.<init>()");}

    void test() {
      System.out.println("TypeC.test()");
      myPrivate();
    }

    //invokeinterface
    public void performMagic() {
      System.out.println("TypeC.performMagic()");
    }

    //invokespecial
    private void myPrivate() {
      System.out.println("TypeC.myPrivate()");
    }
  }

  public static void main(String[] args) {
    final TypeA a = new TypeA();
    final TypeB b = new TypeB();
    final TypeC c = new TypeC();

    callF(a);
    callF(b);
    callPerformMagic(c);

    c.test();

    System.out.print("invokestatic TypeB.g() Expected: 42 Actual: ");
    System.out.println(TypeB.g(39));

    System.out.print("invokestatic TypeC.g() Expected: 16 Actual: ");
    System.out.println(TypeC.g(13));
  }

  @NoInline
  private static void callF(TypeA a) {
    a.f();
  }

  @NoInline
  private static void callPerformMagic(MyInterface myInterface) {
    myInterface.performMagic();
  }
}
