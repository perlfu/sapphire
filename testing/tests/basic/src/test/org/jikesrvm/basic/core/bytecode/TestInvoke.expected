TypeA.<init>()
TypeA.<init>()
TypeB.<init>()
TypeA.<init>()
TypeB.<init>()
TypeC.<init>()
TypeA.f()
TypeB.f()
TypeC.performMagic()
TypeC.test()
TypeC.myPrivate()
invokestatic TypeB.g() Expected: 42 Actual: 42
invokestatic TypeC.g() Expected: 16 Actual: 16
